### ISSUE TITLE
Can we use YOLOv5 data format for segmentation?

### ISSUE THREAD

**Issue by sarmientoj24** [ultralytics/ultralytics#297]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have followed how to create the segmentation from COCO dataset by looking at here https://github.com/ultralytics/yolov5/issues/10621

However, I am getting error like this

```
    data = fetcher.fetch(index)
  File "/home/user/anaconda3/envs/yolo/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 52, in fetch
    return self.collate_fn(data)
  File "/home/user/anaconda3/envs/yolo/lib/python3.7/site-packages/ultralytics/yolo/data/dataset.py", line 172, in collate_fn
    value = torch.stack(value, 0)
TypeError: expected Tensor as element 1 in argument 0, but got tuple
```

My data yaml looks like this
```
path:  # dataset root dir
train: dataset/train  # train images (relative to 'path') 128 images
val: dataset/val  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
names:
  0: object
```

**Comment by Laughing-q** [comment]
@sarmientoj24 yes you can use yolov5 segmentation data format here. it should work.
can you share more error log? also the command please. Thanks!

**Comment by sarmientoj24** [comment]
@Laughing-q 
command is
`yolo task=segment cfg=default.yaml` where my `default.yaml`

has 
```
task: "segment" # choices=['detect', 'segment', 'classify', 'init'] # init is a special case. Specify task to run.
mode: "train" # choices=['train', 'val', 'predict'] # mode to run task in.

# Train settings -------------------------------------------------------------------------------------------------------
model: yolov8m-seg.pt # i.e. yolov8n.pt, yolov8n.yaml. Path to model file
data: data.yaml # i.e. coco128.yaml. Path to data file
```

complete logs
```
val: Scanning /home/user/workspace/proj/test_segmentation/ultralytics/dataset/coco/labels/val.cache... 89 images, 0 backgr
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to test_segmentation/testv118
Starting training for 200 epochs...

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
  0%|          | 0/45 [00:00<?, ?it/s]
Error executing job with overrides: ['cfg=default.yaml']
Traceback (most recent call last):
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/cli.py", line 56, in cli
    func(cfg)
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/hydra/main.py", line 79, in decorated_main
    return task_function(cfg_passthrough)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/v8/segment/train.py", line 152, in train
    model.train(**cfg)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/engine/model.py", line 193, in train
    self.trainer.train()
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/engine/trainer.py", line 185, in train
    self._do_train(int(os.getenv("RANK", -1)), world_size)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/engine/trainer.py", line 283, in _do_train
    for i, batch in pbar:
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/tqdm/std.py", line 1195, in __iter__
    for obj in iterable:
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 521, in __next__
    data = self._next_data()
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1203, in _next_data
    return self._process_data(data)
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1229, in _process_data
    data.reraise()
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/_utils.py", line 434, in reraise
    raise exception
TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 52, in fetch
    return self.collate_fn(data)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/data/dataset.py", line 172, in collate_fn
    value = torch.stack(value, 0)
TypeError: expected Tensor as element 1 in argument 0, but got tuple
```

My folder structure is
```
dataset/
    coco/
        images/
            train/
            val/
        labels/
            train
            val/
```
Example of data
```
0 0.10874 0.708791 0.050813 0.598214 0.015752 0.496566 0.00965447 0.00206044 0.137195 0.00206044 0.139736 0.0254121 0.142785 0.0460165 0.142785 0.0748626 0.139736 0.113324 0.127033 0.128434 0.10874 0.138049 0.0945122 0.147665 0.0990854 0.18544 0.10315 0.21978 0.0945122 0.243132 0.0904472 0.261676 0.0889228 0.292582 0.0919715 0.334478 0.0929878 0.374313 0.0945122 0.4375 0.104675 0.485577 0.114329 0.543269 0.142785 0.628434 0.156504 0.70467
0 0.21189 0.754808 0.165142 0.65728 0.136687 0.546016 0.127033 0.45261 0.121443 0.368819 0.118394 0.344093 0.115854 0.304258 0.114329 0.265797 0.105691 0.236951 0.105691 0.210165 0.10315 0.18544 0.101626 0.168269 0.0990854 0.141484 0.124492 0.135989 0.139736 0.120879 0.145325 0.0975275 0.165142 0.0961538 0.192073 0.0961538 0.219004 0.113324 0.253049 0.133929 0.306402 0.160714 0.311992 0.302198 0.303862 0.326923 0.297764 0.361264 0.299289 0.42239 0.289634 0.502747 0.269817 0.684066 0.269817 0.762363
0 0.840447 0.581044 0.726118 0.495192 0.670732 0.45261 0.624492 0.391484 0.607215 0.37294 0.590447 0.346154 0.57063 0.317308 0.559451 0.280907 0.554878 0.258242 0.557927 0.223901 0.554878 0.177885 0.557927 0.143544 0.634146 0.0865385 0.676321 0.0865385 0.738821 0.0707418 0.815041 0.0748626 0.831809 0.0975275 0.846037 0.12294 0.860264 0.147665 0.855691 0.177198 0.846037 0.208791 0.83435 0.229396 0.829268 0.250687 0.838923 0.273352 0.861789 0.317308 0.934959 0.443681 0.933435 0.586538

```

**Comment by sarmientoj24** [comment]
when batch is one, it goes like this

```
Starting training for 200 epochs...

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
      1/200      1.19G     0.2481     0.7224        inf     0.3141          0        640:  19%|█▊        | 66/353 [00:11<00:
Error executing job with overrides: ['cfg=default.yaml']
Traceback (most recent call last):
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/cli.py", line 56, in cli
    func(cfg)
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/hydra/main.py", line 79, in decorated_main
    return task_function(cfg_passthrough)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/v8/segment/train.py", line 152, in train
    model.train(**cfg)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/engine/model.py", line 193, in train
    self.trainer.train()
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/engine/trainer.py", line 185, in train
    self._do_train(int(os.getenv("RANK", -1)), world_size)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/yolo/engine/trainer.py", line 300, in _do_train
    preds = self.model(batch["img"])
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/nn/tasks.py", line 178, in forward
    return self._forward_once(x, profile, visualize)  # single-scale inference, train
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/nn/tasks.py", line 57, in _forward_once
    x = m(x)  # run
  File "/home/user/anaconda3/envs/yolov5/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user/workspace/proj/test_segmentation/ultralytics/ultralytics/nn/modules.py", line 365, in forward
    return torch.cat(x, self.d)
RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 60 but got size 59 for tensor number 1 in the list.
```

**Comment by Laughing-q** [comment]
@sarmientoj24  can you use the latest version(8.0.5) and try again?
just `pip install --upgrade ultralytics`.

**Comment by sarmientoj24** [comment]
@Laughing-q it is having problems on dataset like this
```
0 0.10874 0.708791 0.050813 0.598214 0.015752 0.496566 0.00965447 0.00206044 0.137195 0.00206044 0.139736 0.0254121 0.142785 0.0460165 0.142785 0.0748626 0.139736 0.113324 0.127033 0.128434 0.10874 0.138049 0.0945122 0.147665 0.0990854 0.18544 0.10315 0.21978 0.0945122 0.243132 0.0904472 0.261676 0.0889228 0.292582 0.0919715 0.334478 0.0929878 0.374313 0.0945122 0.4375 0.104675 0.485577 0.114329 0.543269 0.142785 0.628434 0.156504 0.70467
0 0.21189 0.754808 0.165142 0.65728 0.136687 0.546016 0.127033 0.45261 0.121443 0.368819 0.118394 0.344093 0.115854 0.304258 0.114329 0.265797 0.105691 0.236951 0.105691 0.210165 0.10315 0.18544 0.101626 0.168269 0.0990854 0.141484 0.124492 0.135989 0.139736 0.120879 0.145325 0.0975275 0.165142 0.0961538 0.192073 0.0961538 0.219004 0.113324 0.253049 0.133929 0.306402 0.160714 0.311992 0.302198 0.303862 0.326923 0.297764 0.361264 0.299289 0.42239 0.289634 0.502747 0.269817 0.684066 0.269817 0.762363
0 0.840447 0.581044 0.726118 0.495192 0.670732 0.45261 0.624492 0.391484 0.607215 0.37294 0.590447 0.346154 0.57063 0.317308 0.559451 0.280907 0.554878 0.258242 0.557927 0.223901 0.554878 0.177885 0.557927 0.143544 0.634146 0.0865385 0.676321 0.0865385 0.738821 0.0707418 0.815041 0.0748626 0.831809 0.0975275 0.846037 0.12294 0.860264 0.147665 0.855691 0.177198 0.846037 0.208791 0.83435 0.229396 0.829268 0.250687 0.838923 0.273352 0.861789 0.317308 0.934959 0.443681 0.933435 0.586538
```

**Comment by ichsan2895** [comment]
After One week of debugging. I created a jupyter notebook that mix the code from JSON2COCO & stackoverflow to converting COCO to YOLO and you can download it here : 
https://drive.google.com/file/d/1xhBiWv_Y0HBZQHoWBwF7yjpRrDZhrk4f/view?usp=sharing

Just change the last cell to desired `output_path` and `json_file` path. If you want use bbox annotations, just add `use_segment=False`
```
# the annot will be formatted as bbox, ideal for object detection task
convert_coco_json_to_yolo_txt("yolo_from_Project_2nd", "COCO_Project_2nd/annotations.json", use_segment=False)

# the annot will be formatted as polygon, ideal for instance segmentation task
convert_coco_json_to_yolo_txt("yolo_from_Project_2nd", "COCO_Project_2nd/annotations.json", use_segment=True)
```

**Comment by Laughing-q** [comment]
@sarmientoj24 oh actually both your errors look like image-related issue. can you share some images? or just tell me more info about your images, like resolutions or if it's gray image or something. Thanks!

**Comment by sarmientoj24** [comment]
@Laughing-q image is an XRAY jpeg of one 
![Copy of 000004194](https://user-images.githubusercontent.com/8830319/212484398-1b587ff1-0670-4f4b-ae64-0377a3b86378.jpg)
class (tooth xray)

**Comment by sarmientoj24** [comment]
@ichsan2895 thanks for the suggestion. i was able to convert coco to json using that library already

**Comment by ichsan2895** [comment]
> @ichsan2895 thanks for the suggestion. i was able to convert coco to json using that library already

glad for hearing that.
Are you still got error after using yolo format from my conversion tool?

**Comment by sarmientoj24** [comment]
@ichsan2895 did you change a lot in the original repo? if not, i dont see the reason why to use it over the other.

**Comment by ichsan2895** [comment]
> @ichsan2895 did you change a lot in the original repo? if not, i dont see the reason why to use it over the other. also, i am kind of confused about the file locations in your notebook

the original repo doesn't work in my computer, somehow, the annot.txt didn't create with original repo. Please check my thread in here https://github.com/ultralytics/yolov5/issues/10621

**Comment by sarmientoj24** [comment]
@Laughing-q okay so i just verified that the same dataset worked with YOLOv5 segmentation and not in v8

**Comment by ichsan2895** [comment]
I create a custom segmentation dataset that works in yolov5, yolov7, and
yolov8 from previous notebook that I shared before

Annot format is : class X1, Y1, X2, Y2, .... Xn, Yn

**Comment by sarmientoj24** [comment]
@ichsan2895 what should be the folder structure of the inputs?

**Comment by ichsan2895** [comment]
MyNotebook.ipynb
MyDataset
|->Train
    |-> img_01.jpg
    |-> img_01.txt
|-> Val
    |-> img_02.jpg
    |-> img_02.txt
|-> Test
    |-> img_03.jpg
    |-> img_03.txt
|-> custom_dataset.yaml
```
Make sure the current directory (run `os.getcwd()`) in the directory as same as MyNotebook.ipynb
then run:
```
!yolo task=detect mode=train model=yolov8m.pt data=MyDataset/custom_dataset.yaml epochs=100
```

**Comment by sarmientoj24** [comment]
@ichsan2895 sorry, i meant the folder structure for coco json

**Comment by sarmientoj24** [comment]
@Laughing-q any update on this?

**Comment by ichsan2895** [comment]
> @ichsan2895 sorry, i meant the folder structure for coco json

You can inspect the json format in this file. The COCO json was created by `labelme` annotator
[COCO_Project_1st.zip](https://github.com/ultralytics/ultralytics/files/10443031/COCO_Project_1st.zip)

**Comment by sarmientoj24** [comment]
@Laughing-q any update?

**Comment by sarmientoj24** [comment]
still getting the same problem
```
    return torch.cat(x, self.d)
RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 60 but got size 59 for tensor number 1 in the list.

Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.

```

**Comment by sarmientoj24** [comment]
@ichsan2895 i used your converter and it just gave me the same error

**Comment by Laughing-q** [comment]
@sarmientoj24 hi some users met the same `torch.stack` issue when they set `mosaic!=1.0`.
```bash
    value = torch.stack(value, 0)
TypeError: expected Tensor as element 1 in argument 0, but got tuple
```
the issue could be solved in this PR #699 .

**Comment by hanumanraje** [comment]
> ```
> MyNotebook.ipynb
> MyDataset
> |->Train
>     |-> img_01.jpg
>     |-> img_01.txt
> |-> Val
>     |-> img_02.jpg
>     |-> img_02.txt
> |-> Test
>     |-> img_03.jpg
>     |-> img_03.txt
> |-> custom_dataset.yaml
> ```
> 
> Make sure the current directory (run `os.getcwd()`) in the directory as same as MyNotebook.ipynb then run:
> 
> ```
> !yolo task=detect mode=train model=yolov8m.pt data=MyDataset/custom_dataset.yaml epochs=100
> ```
> 
> I use `yolov8 8.0.5` because AFAIK there are a bug in `yolov8 8.0.6`

Hi can please share notebook, i my case your conversation note book work but model not training give 0 for all loss and gives same image for prediction. I test on your palm dataset as well same result model not training (yolov8)
